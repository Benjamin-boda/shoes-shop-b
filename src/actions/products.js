import database from "../firebase/firebase";
import firebase from 'firebase/app'

//ADD_PRODUCT

export const addProduct = (product) => ({
    type: "ADD_PRODUCT",
    product
});

export const startAddProduct = (productData) => {
    return (dispatch) => {
        const uid = firebase.auth().currentUser.uid;
        const {
            image = "",
            name = "",
            size = "",
            price = 0,
            boughtAt = 0
        } = productData;
        const product = {image, name, size, price, boughtAt};

        return database.ref(`users/${uid}/products`).push(product).then((ref) => {
            dispatch(addProduct({
                id: ref.key,
                ...product
            }))
        })
    }
}

//REMOVE_PRODUCT

export const removeProduct = ({id} = {}) => ({
    type: "REMOVE_PRODUCT",
    id
});

export const startRemoveProduct = ({id}) => {
    return (dispatch) => {
        const uid = firebase.auth().currentUser.uid;
        return database.ref(`users/${uid}/products/${id}`).remove().then(() => {
            dispatch(removeProduct({id}))
        })
    }
}

//EDIT_PRODUCT

export const editProduct = (id, updates) => ({
    type: "EDIT_PRODUCT",
    id,
    updates
});

export const startEditProduct = (id, updates) => {
    return (dispatch) => {
        const uid = firebase.auth().currentUser.uid;
        return database.ref(`users/${uid}/products/${id}`).update(updates).then(() => {
            dispatch(editProduct(id, updates))
        })
    }
}

//SET_PRODUCTS

export const setProducts = (products) => ({
    type: "SET_PRODUCTS",
    products
})

export const startSetProducts = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/products`).once("value").then((snapshot) => {
            const products = [];

            snapshot.forEach((childSnapshot) => {
                products.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
        dispatch(setProducts(products))
        })
    }
}

//REMOVE ALL

export const resetProducts = () => ({
    type: "RESET_PRODUCTS"
});

export const startResetProducts = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        return database.ref(`users/${uid}/products`).remove()
            .then(() => {
                dispatch(resetProducts());
            });
    };
};