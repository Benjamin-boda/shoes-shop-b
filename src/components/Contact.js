import React, { useEffect, useState } from "react";

export const Contact = () => {
    const [subject, setSubject] = useState("")
    const [description, setDescription] = useState("")
    const [submit, setSubmit] = useState("")

    const inputChange = (e) => {
        setSubject(e.target.value)
    }

    const textChange = (e) => {
        setDescription(e.target.value)
    }

    useEffect(() => {
        const request = {
            title: subject,
            object: description
        }
        setSubmit(request)
    }, [subject, description])

    const onSubmit = () => {
        console.log(submit)
    }

    return (
        <div className="contact">
            <div className="contact__header">
                <h1 className="contact__title">Contact us</h1>
                <img className="contact__image" src={require("../images/conctact logo.jpg").default}/>
            </div>
            
            <form className="contact__form" onSubmit={onSubmit}>
                <input data-testid="input" className="contact__form__subject" onChange={inputChange} type="text" placeholder="Subject"/>
                <textarea data-testid="textarea" className="contact__form__description" onChange={textChange} maxLength="200" minLength="10" placeholder="Describe your problem in 200 caracters"/>
                <button className="contact__form__button">Submit</button>
            </form>
        </div>
    )
}