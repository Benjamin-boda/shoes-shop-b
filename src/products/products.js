export const products = [
    {
        name: "shoes 1",
        img: require("../images/shoes1.png"),
        price: 50,
        number: 1
    },
    {
        name: "shoes 2",
        img: require("../images/shoes2.png"),
        price: 50,
        number: 2
    },
    {
        name: "shoes 3",
        img: require("../images/shoes3.png"),
        price: 50,
        number: 3
    },
    {
        name: "shoes 4",
        img: require("../images/shoes4.png"),
        price: 50,
        number: 4
    },
    {
        name: "shoes 5",
        img: require("../images/shoes5.png"),
        price: 50,
        number: 5
    },
    {
        name: "shoes 6",
        img: require("../images/shoes6.png"),
        price: 50,
        number: 6
    },
    {
        name: "shoes 7",
        img: require("../images/shoes7.png"),
        price: 50,
        number: 7
    },
    {
        name: "shoes 8",
        img: require("../images/shoes8.png"),
        price: 50,
        number: 8
    }
]