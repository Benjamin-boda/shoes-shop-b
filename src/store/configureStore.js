import {createStore, combineReducers, compose, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import {authReducer} from "../reducers/auth";
import {ordersReducer} from "../reducers/orders";
import filterReducer from "../reducers/filters";
import productsReducer from "../reducers/product";

// Store creation

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION__COMPOSE__ || compose;

export default () => {
    const store = createStore(
        combineReducers({
            orders: ordersReducer,
            products: productsReducer,
            filters: filterReducer,
            auth: authReducer
        }),
        composeEnhancers(applyMiddleware(thunk))
        // window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
    );
    return store;
};