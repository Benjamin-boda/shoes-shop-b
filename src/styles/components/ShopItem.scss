.shopitem__content {
    height: 85vh;
    width: 100%;
}

.shopitem {
    align-items: center;
    display: flex;
    flex-direction: column;
    justify-content: space-around;
    margin: 0 auto;
    max-width: 80rem;
    padding: $l-size 0 0 0;

    @media (min-width: $desktop-breakpoint) {
        flex-direction: row;
    }
}

.shopitem__border {
    border: black .2rem solid;
    width: 100%;
}

.shopitem__image {
    -moz-box-shadow: 10px 10px 5px 0px #656565;
    -webkit-box-shadow: 10px 10px 5px 0px #656565;
    -o-box-shadow: 10px 10px 5px 0px #656565;
    box-shadow: 10px 10px 5px 0px #656565;
    background: white;
    border-radius: .5rem;
    width: 20rem;

    @media (min-width: $desktop-breakpoint) {
        width: 30rem;
    }
    
}

.shopitem__name {
    font-family: "MonserratBold";
    font-size: $font-size-medium;
    text-align: center;
}

.shopitem__select {
    border-radius: .5rem;
    font-family: "MonserratRegular";
    font-size: $font-size-small;
    padding: $xs-size;
}

.shopitem__size {
    align-items: center;
    display: flex;
    justify-content: space-between;
}

.shopitem__size__text {
    font-size: $font-size-small;
    font-family: "MonserratRegular";
    margin: 0 $m-size 0 0;
}

.shopitem__button {
    -moz-box-shadow: 10px 10px 5px 0px #656565;
    -webkit-box-shadow: 10px 10px 5px 0px #656565;
    -o-box-shadow: 10px 10px 5px 0px #656565;
    box-shadow: 10px 10px 5px 0px #656565;
    background: grey;
    border: none;
    border-radius: .5rem;
    color: white;
    font-family: "MonserratBold";
    font-size: $font-size-small;
    margin: $xl-size;
    padding: $xs-size;
}