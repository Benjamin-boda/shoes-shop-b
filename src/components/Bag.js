import React, { useEffect, useState } from "react";
import {useSelector, useDispatch} from "react-redux";
import {startRemoveProduct, startEditProduct, startResetProducts} from "../actions/products";
import {startAddOrder} from "../actions/orders";
import products from "../tests/fixtures/products";

export const Bag = () => {
    const [sizeUpdate, setSizeUpdate] = useState("")
    const [productUpdate, setProductUpdate] = useState("")
    const productsInStore  = useSelector(state => state.products)
    const dispatch = useDispatch()

    const sizes = [37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47]
    

    const sizeEdit = (e) => {
        setSizeUpdate(e.target.value)
    }

    useEffect(() => {
        if (productUpdate !== "" & productsInStore.lenght > 0) {
            const filterProductArray = productsInStore.filter((product) => productUpdate === product.name)
            const productEdit = {
                image : {
                    default : filterProductArray[0].image.default
                },
                name : filterProductArray[0].name,
                size : sizeUpdate,
                price : filterProductArray[0].price,
                boughtAt : filterProductArray[0].boughtAt
            }
            const id = filterProductArray[0].id
            dispatch(startEditProduct(id, productEdit))    
        }
    }, [sizeUpdate])

    const paymentOrder = () => {
        productsInStore.map((product) => {
            product.boughtAt = Date()
            dispatch(startAddOrder(product))})
        dispatch(startResetProducts())              
    }

    const sumPrice = productsInStore.length > 0 ? productsInStore.map(product => product.price).reduce((acc, price) => acc + price) : "0"

    return (
        <div className="bag__content">
            <div className="bag__box">
                {productsInStore.length > 0 ?
                    productsInStore.map((product) => 
                        <div className="bag" key={product.name} onMouseEnter={() => setProductUpdate(product.name)}>
                            <img className="bag__image" src={product.image.default}/>
                            <p data-testid={product.name} className="bag__name">{product.name}</p>
                            <select className="bag__select" name="size" onChange={sizeEdit}>
                                <option value={product.size}>{product.size}</option>
                                {sizes.map((size) => 
                                    <option value={size} key={size}>{size}</option>
                                    )}
                            </select>
                            <p className="bag__name">Price : {product.price} €</p>
                            <button className="bag__button" onClick={() => dispatch(startRemoveProduct(product))}><img className="bag__cross__button" src={require("../images/cross button.png").default}/></button>
                        </div>
                    ) : <p className="bag__empty">Nothing in the bag</p>
                }
            </div>
            
            <p className="bag__price">Total price : {sumPrice} €</p>

            <div className="bag__payment__div">
                <button className="bag__payment" onClick={paymentOrder} disabled={productsInStore.length === 0 ? true : false}>Payment</button>
            </div>
            
        </div>
    )
}