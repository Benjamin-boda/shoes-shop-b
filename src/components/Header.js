import React, {useState ,useContext} from "react";
import {Link} from "react-router-dom";
import {useDispatch, useSelector} from "react-redux";
import {startLogin, startLogout} from "../actions/auth";
import {AuthContext} from "../firebase/Auth";
import {LoginModal} from "./LoginModal";

export const Header = () => {
    const {currentUser} = useContext(AuthContext);
    const [accountHovering, setAccountHovering] = useState(false)
    const [modalIsOpen, setModalIsOpen] = useState(false)
    const productsInStore = useSelector(state => state.products)
    const dispatch = useDispatch()

    const openModal = () => {
        if (!!currentUser) {
            setModalIsOpen(false)
        } else {
            setModalIsOpen(true)
        }
    }

    return (
        <div className="header">
            <img className="header__logo" src={require("../images/logo.png").default}/>
            <Link className="header__link" to="/">Home</Link>
            <Link className="header__link" to={!!currentUser ? "/shop/1" : ""} onClick={openModal}>Shop</Link>
            <Link className="header__link" to="/contact">Contact</Link>
            <Link className="header__link" to="/help">Help</Link>
            
            <div className="header__account">
                <Link 
                    data-testid="account"
                    className="header__link__account"
                    to=""
                    onClick={() => !!currentUser ? undefined : dispatch(startLogin())}
                    onMouseEnter={() => !!currentUser ? setAccountHovering(true) : setAccountHovering(false)}
                >
                    {!!currentUser ? "User account" : "Login"}
                </Link>
                {accountHovering ? 
                    <ul className="header__dropdown" onClick={() => setAccountHovering(false)} onMouseLeave={() => setAccountHovering(false)}>
                        <Link className="header__link__account__dropdown" to="/orders">
                            <li className="header__dropdown__item">My account</li>
                        </Link>
                        <li className="header__dropdown__item" onClick={() => dispatch(startLogout())}>Logout</li>
                        <li className="header__dropdown__item" onClick={() => setAccountHovering(false)}>Close</li>
                    </ul> : undefined
                }
            </div>
            
            <div>
                <Link to={!!currentUser ? "/bag" : ""} onClick={openModal}><img className="header__bag" src={require("../images/bag.png").default}/></Link>
                <p className="header__bag__counter">{productsInStore.length > 0 ? productsInStore.length : 0}</p>
            </div>
            

            <LoginModal modalIsOpen={modalIsOpen} setModalIsOpen={setModalIsOpen}/>
        </div>
    )
}