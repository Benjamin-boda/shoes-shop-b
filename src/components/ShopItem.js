import React, {useState} from "react";
import { useParams } from "react-router-dom";
import {useDispatch} from "react-redux";
import {products} from "../products/products";
import {startAddProduct} from "../actions/products";

export const ShopItem = () => {
    const [sizeChoice, setSizeChoice] = useState("")
    const [addClicked, setAddClicked] = useState(false)
    const sizes = [37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47]
    const {productnumber} = useParams()
    const dispatch = useDispatch()

    const filterProduct = products.filter((product) => product.number === parseInt(productnumber))
    const userProduct = {
            image : filterProduct[0].img,
            name : filterProduct[0].name,
            size : sizeChoice,
            price : filterProduct[0].price,
            boughtAt : 0
    }

    const sizeOnChange = (e) => {
        setSizeChoice(e.target.value)
    }

    return (
        <div className={`shopitem__content shop__shoes__${Math.floor(Math.random() * 4)}`}>
            <div className="shopitem">
                <div>
                    {products.map((product) => 
                        product.number === parseInt(productnumber) ? 
                        <div key={product.number}>
                            <img className="shopitem__image" src={product.img.default}/>
                            <p className="shopitem__name">{product.name}</p>
                        </div> : undefined
                        )}
                </div>
                
                <div>
                    <div className="shopitem__size">
                        <p className="shopitem__size__text">Size : </p>
                        <select className="shopitem__select" name="size" onChange={sizeOnChange}>
                            <option value="">Select your size</option>
                            {sizes.map((size) => 
                                <option value={size} key={size}>{size}</option>
                                )}
                        </select>
                    </div>
                    
                    <button data-testid="button" className="shopitem__button" onClick={() => sizeChoice !== "" ? 
                        dispatch(startAddProduct(userProduct))
                        : setAddClicked(true)
                    }>
                        Add to bag
                    </button>
                    {sizeChoice === "" & addClicked ? <p>You need to select a size first</p> : undefined}
                </div>
            </div>
        </div>
        
    )
}