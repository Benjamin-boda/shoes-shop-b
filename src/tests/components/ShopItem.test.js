import React from "react";
import {ShopItem} from "../../components/ShopItem";
import {render, cleanup, fireEvent} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect';
import products from "../fixtures/products";
import { MemoryRouter } from 'react-router-dom';
import ReactRouter from 'react-router'
import { useSelector } from 'react-redux';
import configureStore from "redux-mock-store";
import {productsReducer} from "../../reducers/product";
import {AuthProvider} from "../../firebase/Auth";

afterEach(cleanup);

jest.mock('react-redux', () => ({
    useDispatch: jest.fn(),
    useSelector: jest.fn()
}));

jest.spyOn(ReactRouter, 'useParams').mockReturnValue({ productnumber: '1' });

useSelector.mockImplementation((selector) => selector({
    products : products
}))

const renderWithRedux = (ui, { initialState, ...renderOptions } = {}) => {
    const mockStore = configureStore(productsReducer, initialState);
    const store = mockStore({});
    const Wrapper = ({ children }) => (
        <MemoryRouter>
            <AuthProvider store={store}>{children}</AuthProvider>
        </MemoryRouter>
    );
    return render(ui, { wrapper: Wrapper, ...renderOptions });
  };

const { asFragment, getByText, getByTestId } = renderWithRedux(<ShopItem products={products}/>)

test("should match the snapshot", () => {
    
    expect(asFragment(<ShopItem/>)).toMatchSnapshot()
})

test("should contain this text", () => {
    jest.spyOn(ReactRouter, 'useParams').mockReturnValue({ productnumber: '1' });

    renderWithRedux(<ShopItem products={products}/>)

    expect(getByText("shoes 1")).toBeTruthy()
})

test("should contain this text", () => {
    jest.spyOn(ReactRouter, 'useParams').mockReturnValue({ productnumber: '1' });

    renderWithRedux(<ShopItem products={products}/>)

    fireEvent.click(getByTestId("button"))
    expect(getByText("You need to select a size first")).toBeTruthy()
})