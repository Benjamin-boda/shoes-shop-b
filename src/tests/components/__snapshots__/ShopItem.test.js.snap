// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="shopitem__content shop__shoes__3"
  >
    <div
      class="shopitem"
    >
      <div>
        <div>
          <img
            class="shopitem__image"
          />
          <p
            class="shopitem__name"
          >
            shoes 1
          </p>
        </div>
      </div>
      <div>
        <div
          class="shopitem__size"
        >
          <p
            class="shopitem__size__text"
          >
            Size : 
          </p>
          <select
            class="shopitem__select"
            name="size"
          >
            <option
              value=""
            >
              Select your size
            </option>
            <option
              value="37"
            >
              37
            </option>
            <option
              value="38"
            >
              38
            </option>
            <option
              value="39"
            >
              39
            </option>
            <option
              value="40"
            >
              40
            </option>
            <option
              value="41"
            >
              41
            </option>
            <option
              value="42"
            >
              42
            </option>
            <option
              value="43"
            >
              43
            </option>
            <option
              value="44"
            >
              44
            </option>
            <option
              value="45"
            >
              45
            </option>
            <option
              value="46"
            >
              46
            </option>
            <option
              value="47"
            >
              47
            </option>
          </select>
        </div>
        <button
          class="shopitem__button"
          data-testid="button"
        >
          Add to bag
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
